runner/full.go

```diff
@@ -88,29 +89,37 @@ func processServer(ctx context.Context, server *state.Server, globalCollectionOp
 	var newState state.PersistedState
 	var collectionStatus state.CollectionStatus
 	var err error
+	newGrant.Config = pganalyze_collector.ServerMessage_Config{Features: &pganalyze_collector.ServerMessage_Features{}}
+
+	if server.Pause.Load().Pause {
+		logger.PrintWarning("Snapshot processing disabled by pganalyze server: %s", server.Pause.Load().Reason)
+		return newState, newGrant, collectionStatus, nil
+	}
 
 	err = checkReplicaCollectionDisabled(ctx, server, globalCollectionOpts, logger)
 	if err != nil {
-		return state.PersistedState{}, state.Grant{}, state.CollectionStatus{}, err
+		return newState, newGrant, collectionStatus, err
 	}
 
-	if !globalCollectionOpts.ForceEmptyGrant {
-		// Note: In case of server errors, we should reuse the old grant if its still recent (i.e. less than 50 minutes ago)
+	if server.WebSocket.Load() != nil {
+		newGrant = *server.Grant.Load()
+	} else if !globalCollectionOpts.ForceEmptyGrant {
 		newGrant, err = grant.GetDefaultGrant(ctx, server, globalCollectionOpts, logger)
 		if err != nil {
-			if server.Grant.Valid {
+			if server.Grant.Load().Valid {
 				logger.PrintVerbose("Could not acquire snapshot grant, reusing previous grant: %s", err)
 			} else {
-				return state.PersistedState{}, state.Grant{}, state.CollectionStatus{}, err
+				return newState, newGrant, collectionStatus, err
 			}
 		} else {
-			server.Grant = newGrant
+			server.Grant.Store(&newGrant)
```

Comment:
It seems like we actually do this twice, once here, and then once in the caller of this function. I think we should be good to drop this here? (or in the parent function?)